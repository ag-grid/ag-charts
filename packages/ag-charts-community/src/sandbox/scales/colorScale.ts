import { clamp } from '../../util/number';
import { type OKLCH, oklchToRgba, rgbaToOklch, stringToRgba } from '../util/colors.util';
import { interpolateColor } from '../util/interpolate.util';
import { BaseScale } from './baseScale';

export class ColorScale extends BaseScale<number, string> {
    private readonly oklchRange: OKLCH[];

    constructor(domain: number[], range: string[]) {
        if (domain.length < 2) {
            throw new Error('Color scale domain must contain at least 2 values.');
        }
        if (domain.length !== range.length) {
            throw new Error('Color scale range size must match the provided domain size.');
        }
        super(domain, range);
        this.oklchRange = range.map((value) => {
            const { r, g, b, a } = stringToRgba(value);
            return rgbaToOklch(r, g, b, a);
        });
    }

    convert(value: number) {
        const { domain, oklchRange } = this;
        const i = this.findBoundingDomainIndex(value);
        const d = clamp(0, (value - domain[i]) / (domain[i + 1] - domain[i]), 1);
        const interpolator = interpolateColor(oklchRange[i], oklchRange[i + 1]);
        const { l, c, h, a } = interpolator(d);
        const { r, g, b } = oklchToRgba(l, c, h, a);
        return `rgba(${r},${g},${b},${a})`;
    }

    override invert(_value: string) {
        return 0;
    }

    private findBoundingDomainIndex(value: number) {
        for (let i = 1; i < this.domain.length; i++) {
            if (value >= this.domain[i - 1] && value <= this.domain[i]) {
                return i - 1;
            }
        }
        return -1;
    }
}
