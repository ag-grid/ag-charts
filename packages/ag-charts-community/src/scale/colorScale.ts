import { Scale } from './scale';
import { Color } from '../util/color';
import interpolateColor from '../interpolate/color';

export class ColorScale implements Scale<number, string, number> {
    domain = [0, 1];

    private _range = ['red', 'green'];
    private parsedRange = this._range.map((v) => Color.fromString(v));
    get range() {
        return this._range;
    }
    set range(values: string[]) {
        this._range = values;
        this.parsedRange = values.map((v) => Color.fromString(v));
    }

    convert(x: number) {
        const { domain, range, parsedRange } = this;
        const d0 = domain[0];
        const d1 = domain[domain.length - 1];
        const r0 = range[0];
        const r1 = range[range.length - 1];

        if (x <= d0) {
            return r0;
        }

        if (x >= d1) {
            return r1;
        }

        let index: number;
        let q: number;

        if (domain.length === 2) {
            const t = (x - d0) / (d1 - d0);
            const step = 1 / (range.length - 1);
            index = range.length <= 2 ? 0 : Math.min(Math.floor(t * (range.length - 1)), range.length - 2);
            q = (t - index * step) / step;
        } else {
            for (index = 0; index < domain.length - 2; index++) {
                if (x < domain[index + 1]) {
                    break;
                }
            }
            const a = domain[index];
            const b = domain[index + 1];
            q = (x - a) / (b - a);
        }

        const c0 = parsedRange[index];
        const c1 = parsedRange[index + 1];
        return interpolateColor(c0, c1)(q);
    }
}
