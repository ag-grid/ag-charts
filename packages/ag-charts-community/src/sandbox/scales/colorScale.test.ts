import { ColorScale } from './colorScale';

describe('ColorScale', () => {
    it('should throw an error if the domain has less than 2 values', () => {
        expect(() => new ColorScale([1], ['#ff0000'])).toThrow('Color scale domain must contain at least 2 values.');
    });

    it('should throw an error if the domain size does not match the range size', () => {
        expect(() => new ColorScale([1, 2, 3], ['#ff0000', '#00ff00'])).toThrow(
            'Color scale range size must match the provided domain size.'
        );
    });

    it('should convert domain values to corresponding color values', () => {
        const scale = new ColorScale([1, 2, 3], ['#ff0000', '#00ff00', '#0000ff']);

        expect(scale.convert(1)).toBe('rgba(255,0,0,1)');
        expect(scale.convert(2)).toBe('rgba(0,255,0,1)');
    });

    it('should interpolate colors correctly within the domain range', () => {
        const scale = new ColorScale([1, 2, 3], ['#ff0000', '#00ff00', '#0000ff']);

        expect([1, 1.25, 1.5, 1.75, 2, 2.1, 2.31, 2.57, 2.83, 2.99, 3].map((n) => scale.convert(n))).toMatchSnapshot();
    });
});
