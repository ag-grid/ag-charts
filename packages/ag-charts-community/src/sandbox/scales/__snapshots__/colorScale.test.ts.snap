// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorScale should interpolate colors correctly within the domain range 1`] = `
[
  "rgba(255,0,0,1)",
  "rgba(255,86,0,1)",
  "rgba(249,149,0,1)",
  "rgba(194,206,0,1)",
  "rgba(0,255,0,1)",
  "rgba(0,248,98,1)",
  "rgba(0,223,178,1)",
  "rgba(0,171,241,1)",
  "rgba(0,96,255,1)",
  "rgba(0,15,255,1)",
  "rgba(0,0,255,1)",
]
`;
